### PR TITLE
Bug 1470953 - Refactor TabTrayViewController to use TabDisplayManager

### PR DESCRIPTION
**Pull request by farhanpatel**
This creates a new `TabDisplayer` protocol that both TabTray and TopTabs implement.

### CODE REVIEW

File: Client/Frontend/Browser/TabTrayController.swift [hunk 1]
```diff
@@ -1175,3 +639,223 @@ class TrayToolbar: UIView {
         maskButton.applyTheme(theme)
     }
 }
+
```
Comment (farhanpatel) [inline Client/Frontend/Browser/TabTrayController.swift]: All this was simply moved from the top to the bottom of the file.

File: Client/Frontend/Browser/TabTrayController+KeyCommands.swift [hunk 2]
```diff
@@ -77,7 +77,8 @@ extension TabTrayController {
         }
 
         let nextIndex = max(0, min(currentIndex + step, tabs.count - 1))
-        let nextTab = tabs[nextIndex]
-        tabManager.selectTab(nextTab)
+        if let nextTab = tabs[safe: nextIndex] {
```
Comment (justindarc) [inline Client/Frontend/Browser/TabTrayController+KeyCommands.swift]: 👍

File: Client/Frontend/Browser/TabTrayController.swift [hunk 3]
```diff
@@ -497,50 +251,69 @@ class TabTrayController: UIViewController {
...
+        return tabCell
+    }
+}
+
+extension TabTrayController {
+
+    @objc func didTapLearnMore() {
+        let appVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
+        if let langID = Locale.preferredLanguages.first {
+            let learnMoreRequest = URLRequest(url: "https://support.mozilla.org/1/mobile/\(appVersion ?? "0.0")/iOS/\(langID)/private-browsing-ios".asURL!)
```
Comment (justindarc) [inline Client/Frontend/Browser/TabTrayController.swift]: I know it was like this before, but I wonder if we should extract this out somewhere like a string. Maybe not though.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 4]
```diff
@@ -575,120 +349,28 @@ extension TabTrayController: PresentingModalViewControllerDelegate {
...
+        bounds = bounds.offsetBy(dx: collectionView.contentInset.left, dy: collectionView.contentInset.top)
+        bounds.size.width -= collectionView.contentInset.left + collectionView.contentInset.right
+        bounds.size.height -= collectionView.contentInset.top + collectionView.contentInset.bottom
+        // visible cells do sometimes return also not visible cells when attempting to go past the last cell with VoiceOver right-flick gesture; so make sure we have only visible cells (yeah...)
+        visibleCells = visibleCells.filter { !$0.frame.intersection(bounds).isEmpty }
 
-        // Only redraw if there is more than one tab in the tray.
-        guard tabs.count > 1 else {
-            return
+        let cells = visibleCells.map { self.collectionView.indexPath(for: $0)! }
```
Comment (justindarc) [inline Client/Frontend/Browser/TabTrayController.swift]: There's no reasonable chance of this failing, is there?

File: Client/Frontend/Browser/TabTrayController.swift [hunk 5]
```diff
@@ -575,120 +349,28 @@ extension TabTrayController: PresentingModalViewControllerDelegate {
...
+        let indexPaths = cells.sorted { (a: IndexPath, b: IndexPath) -> Bool in
+            return a.section < b.section || (a.section == b.section && a.row < b.row)
         }
 
-        let updated = [ selected, previous ]
-            .compactMap { $0 }
-            .compactMap { tabs.index(of: $0) }
-            .map { IndexPath(item: $0, section: 0) }
+        if indexPaths.count == 0 {
```
Comment (justindarc) [inline Client/Frontend/Browser/TabTrayController.swift]: Should this be `guard indexPaths.count > 0 else { return ... }`? potato/potatoe

File: Client/Frontend/Browser/TabTrayController.swift (hunk 3 above)
Comment (farhanpatel) [reply Client/Frontend/Browser/TabTrayController.swift]: If this was a simple static string I would. But we use the appVersion and locale to build this URL. I think it makes sense to just leave it here.

File: Client/Frontend/Browser/TabTrayController.swift (hunk 4 above)
Comment (farhanpatel) [reply Client/Frontend/Browser/TabTrayController.swift]: visibleCells is an array of cells on the collectionview. and then this map just looks for the indexpath of those cells. Dont see this crashing.

**Pull request merged**